dotnet/src/Azure.Iot.Operations.Protocol/CloudEvent.cs

```diff
@@ -27,21 +27,75 @@ public class CloudEvent(Uri source, string type = "", string specversion = "1.0"
         /// This enables the interpretation of the context.
         /// Compliant event producers MUST use a value of 1.0 when referring to this version of the specification.
         /// </summary>
-        public string SpecVersion => specversion;
+        public string SpecVersion
+        {
+            get
+            {
+                return _specVersion;
+            }
+            set
+            {
+                if (string.IsNullOrWhiteSpace(value))
+                {
+                    throw new ArgumentException("SpecVersion must not be an empty string");
+                }
+
+                _specVersion = value;
+            }
+        }
+
+        private string _specVersion = specversion;
 
         /// <summary>
         /// Contains a value describing the type of event related to the originating occurrence.
         /// Often this attribute is used for routing, observability, policy enforcement, etc.
         /// The format of this is producer defined and might include information such as the version of the type
         /// </summary>
-        public string Type { get; set; } = type;
+        /// <remarks>
+        /// This value cannot be null or whitespace
+        /// </remarks>
+        public string Type
+        {
+            get
+            {
+                return _type;
+            }
+            set
+            {
+                if (string.IsNullOrWhiteSpace(value))
+                {
+                    throw new ArgumentException("The type cannot be null or whitespace");
+                }
+
+                _type = value;
+            }
+        }
+
+        private string _type = type;
 
         /// <summary>
         ///  Identifies the event. Producers MUST ensure that source + id is unique for each distinct event.
         ///  If a duplicate event is re-sent (e.g. due to a network error) it MAY have the same id.
         ///  Consumers MAY assume that Events with identical source and id are duplicates.
         /// </summary>
-        public string? Id { get; set; } = null!; // although id is required, we want update it in the same instance from the sender.
+        public string? Id
+        {
+            get
+            {
+                return _id;
+            }
+            set
+            {
+                if (string.IsNullOrWhiteSpace(value))
+                {
+                    throw new ArgumentException("Id cannot be null or whitespace");
+                }
+
+                _id = value;
+            }
+        }
+
+        private string _id = Guid.NewGuid().ToString();
```

Comment:
Now that this default is here like this, does Id still need to be optional, or can it be required for extra safety?

Reply:
I'd like to make it required, but that would cause a breaking change at this point